File: app/src/main/java/uz/kidzone/app/kidzo/ContentFilter.java
package uz.kidzone.app.kidzo;

import android.content.Context;
import androidx.annotation.Nullable;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class ContentFilter {

    private final List<ContentItem> stories = new ArrayList<>();
    private final List<ContentItem> songs   = new ArrayList<>();

    /** Test and production use — accepts JSON string. */
    public ContentFilter(String contentJson) throws JSONException {
        JSONObject root = new JSONObject(contentJson);
        parseItems(root.optJSONArray("stories"), stories);
        parseItems(root.optJSONArray("songs"),   songs);
    }

    /** Production: reads from assets/www/content.json */
    public static ContentFilter fromAssets(Context ctx) throws IOException, JSONException {
        try (InputStream is = ctx.getAssets().open("www/content.json");
             BufferedReader reader = new BufferedReader(
                     new InputStreamReader(is, java.nio.charset.StandardCharsets.UTF_8))) {
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) sb.append(line);
            return new ContentFilter(sb.toString());
        }
    }

    private void parseItems(JSONArray arr, List<ContentItem> out) throws JSONException {
        if (arr == null) return;
        for (int i = 0; i < arr.length(); i++) {
            JSONObject o = arr.getJSONObject(i);
            JSONObject title = o.optJSONObject("title");
            out.add(new ContentItem(
                o.getString("id"),
                o.optString("emoji", ""),
                title != null ? title.optString("uz", "") : "",
                title != null ? title.optString("ru", "") : "",
                title != null ? title.optString("en", "") : "",
                o.optString("category", "")
            ));
        }
    }

    /** Top 5 stories (songs removed from app). */
    public List<ContentItem> getTop5() {
        int limit = Math.min(5, stories.size());
        return new ArrayList<>(stories.subList(0, limit));
    }

    /** Filter stories by keyword — title (uz/ru/en) or category. */
    public List<ContentItem> getFiltered(String query) {
        if (query == null || query.trim().isEmpty()) return getTop5();
        String q = query.toLowerCase().trim();
        List<ContentItem> result = new ArrayList<>();
        for (ContentItem item : stories) { if (matches(item, q)) result.add(item); }
        int limit = Math.min(5, result.size());
        return new ArrayList<>(result.subList(0, limit));
    }

    @Nullable
    public ContentItem findById(@Nullable String id) {
        if (id == null) return null;
        for (ContentItem item : stories) { if (item.id.equals(id)) return item; }
        for (ContentItem item : songs)   { if (item.id.equals(id)) return item; }
        return null;
    }

    /** Format a list of items as a Gemini prompt block. */
    public String toPromptBlock(List<ContentItem> items) {
        StringBuilder sb = new StringBuilder();
        for (ContentItem item : items) sb.append(item.toPromptLine()).append("\n");
        return sb.toString().trim();
    }

    private boolean matches(ContentItem item, String q) {
        return item.titleUz.toLowerCase().contains(q)
            || item.titleRu.toLowerCase().contains(q)
            || item.titleEn.toLowerCase().contains(q)
            || item.category.toLowerCase().contains(q);
    }
}

File: app/src/androidMain/kotlin/uz/kidzone/app/kidzo/ContentFilter.kt
package uz.kidzone.app.kidzo

import android.content.Context
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets

class ContentFilter @Throws(JSONException::class) constructor(contentJson: String) {

    private val stories = mutableListOf<ContentItem>()
    private val songs   = mutableListOf<ContentItem>()

    init {
        val root = JSONObject(contentJson)
        parseItems(root.optJSONArray("stories"), stories)
        parseItems(root.optJSONArray("songs"),   songs)
    }

    @Throws(JSONException::class)
    private fun parseItems(arr: JSONArray?, out: MutableList<ContentItem>) {
        if (arr == null) return
        for (i in 0 until arr.length()) {
            val o = arr.getJSONObject(i)
            val title = o.optJSONObject("title")
            out.add(
                ContentItem(
                    id       = o.getString("id"),
                    emoji    = o.optString("emoji", ""),
                    titleUz  = title?.optString("uz", "") ?: "",
                    titleRu  = title?.optString("ru", "") ?: "",
                    titleEn  = title?.optString("en", "") ?: "",
                    category = o.optString("category", ""),
                )
            )
        }
    }

    /** Top 5 stories (songs removed from app). */
    fun getTop5(): List<ContentItem> {
        val limit = minOf(5, stories.size)
        return stories.subList(0, limit).toList()
    }

    /** Filter stories by keyword — title (uz/ru/en) or category. */
    fun getFiltered(query: String?): List<ContentItem> {
        if (query.isNullOrBlank()) return getTop5()
        val q = query.lowercase().trim()
        val result = stories.filter { matches(it, q) }
        val limit = minOf(5, result.size)
        return result.subList(0, limit)
    }

    fun findById(id: String?): ContentItem? {
        if (id == null) return null
        return stories.find { it.id == id } ?: songs.find { it.id == id }
    }

    /** Format a list of items as a Gemini prompt block. */
    fun toPromptBlock(items: List<ContentItem>): String {
        return items.joinToString("\n") { it.toPromptLine() }
    }

    private fun matches(item: ContentItem, q: String): Boolean {
        return item.titleUz.lowercase().contains(q)
            || item.titleRu.lowercase().contains(q)
            || item.titleEn.lowercase().contains(q)
            || item.category.lowercase().contains(q)
    }

    companion object {
        /** Production: reads from assets/www/content.json */
        @JvmStatic
        @Throws(IOException::class, JSONException::class)
        fun fromAssets(ctx: Context): ContentFilter {
            ctx.assets.open("www/content.json").use { inputStream ->
                val reader = BufferedReader(InputStreamReader(inputStream, StandardCharsets.UTF_8))
                val sb = StringBuilder()
                var line = reader.readLine()
                while (line != null) {
                    sb.append(line)
                    line = reader.readLine()
                }
                return ContentFilter(sb.toString())
            }
        }
    }
}
